Added PEP 442 implementation and test case

ax487 opened the pull request:
This should adress #3612. I added a simple test case to ensure that the `__del__` methods is called when the cdef class is released.

da-woods commented:
It seems there's a reasonable amount of (potential) complexity here that isn't being tested:
* class with a `__dealloc__` method and a `__del__` method.
* Various combinations of inheritance.
* Does a scheme where an object resurrects itself in `__del__` work correctly?
* Does `__del__` swallow exceptions like it's supposed to?

ax487 commented:
Sorry, first time contributor...

I added some simple tests, finding that `__dealloc__` is *not* called when `__del__` is present. I guess that necessitates some edits at `ModuleNode.generate_dealloc_function`?

scoder commented:
> I added some simple tests, finding that `__dealloc__` is *not* called when `__del__` is present.

It's just that your tests accidentally call the wrong class.

ax487 commented:
OK, I fixed the error and added more tests. There is one failing test, namely an object resurrecting itself, based on this [example](https://en.wikipedia.org/wiki/Object_resurrection). The `__del__` method is called only once as expected, but apparently the object is never `__dealloc__`ated, which seems like an error?!

da-woods commented:
> OK, I fixed the error and added more tests. There is one failing test, namely an object resurrecting itself, based on this [example](https://en.wikipedia.org/wiki/Object_resurrection). The `__del__` method is called only once as expected, but apparently the object is never `__dealloc__`ated, which seems like an error?!

In `test_zombie_object` I don't think `__dealloc__` should be called while a reference is still held as the `global c` (at least within the timeline of the function). Maybe something like

```
global c
print("start")
i = immortal()
i = None
import gc
gc.collect()
print("del c")
del c  # __dealloc__ should be called here
gc.collect()  # or possibly here
print("finish")
```

I think most of this should just work right (because it's handled by the Python interpreter) but it's good to make sure.

ax487 commented:
I added a test case `test_del_with_exception` which throws an exception in `__del__`. It seems like this exception is indeed swallowed.

@da-woods: I adapted the resurrection test as you suggested, but there is still no `__dealloc__` called even as the global reference is deleted...

da-woods commented:
Hmmmm... My understanding is that `__del__` is only ever supposed to be called once (so it should be resurrected the first time, but not the second). It's possible that I'm misunderstanding it though...

da-woods commented:
Right - got it I think:

If the object isn't a "GC class" (i.e. `cython.no_gc` is set, or it just doesn't have an object members like your `immortal` class) then `__del__` can be called multiple times. In this case the object keeps getting resurrected so it's never possible to deallocate it (as is happening here).

If the object is a "GC class" (e.g. if it has a `cdef object x` member) then `__del__` can only be called once. And so `__dealloc__` is called after `del c`.

It looks like this is all working correctly

ax487 commented:
OK, I added the test with the predicted results...

da-woods commented:
There's definitely an issue with exceptions:

```
cdef class C:
    def __dealloc__(self):
        print("".join(["a", "b", "c"])) # some python code

def f():
    c = C()
    raise RuntimeError()

cdef class D:
    def __del__(self):
        print("".join(["a", "b", "c"])) # some python code

def g():
    c = D()
    raise RuntimeError()
```

In both cases (for classes `C` and `D`) the `__dealloc__`/`__del__` method is called while an exception is set. We handle this with `PyErr_Fetch` and `PyErr_Restore` for the `__dealloc__` case:

```
static void __pyx_tp_dealloc_11testdealloc_C(PyObject *o) {
  #if CYTHON_USE_TP_FINALIZE
  if (unlikely((PY_VERSION_HEX >= 0x03080000 || __Pyx_PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE)) && Py_TYPE(o)->tp_finalize) && (!PyType_IS_GC(Py_TYPE(o)) || !_PyGC_FINALIZED(o))) {
    if (PyObject_CallFinalizerFromDealloc(o)) return;
  }
  #endif
  {
    PyObject *etype, *eval, *etb;
    PyErr_Fetch(&etype, &eval, &etb);
    __Pyx_SET_REFCNT(o, Py_REFCNT(o) + 1);
    __pyx_pw_11testdealloc_1C_1__dealloc__(o);
    __Pyx_SET_REFCNT(o, Py_REFCNT(o) - 1);
    PyErr_Restore(etype, eval, etb);
  }
  (*Py_TYPE(o)->tp_free)(o);
}
```

However for the `__del__` case we don't handle it resulting in

```
>>> testdealloc.g()
Traceback (most recent call last):
  File "testdealloc.pyx", line 15, in testdealloc.g
    raise RuntimeError()
RuntimeError
Exception ignored in: 'testdealloc.D.__del__'
Traceback (most recent call last):
  File "testdealloc.pyx", line 15, in testdealloc.g
    raise RuntimeError()
RuntimeError: 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <cyfunction g at 0x7f0a0a331930> returned NULL without setting an error
```

I think the solution is to create a small wrapper as like for `__dealloc__` which essentially only does the fetch and restore. However, I'm not immediately sure where that's generated.

da-woods commented:
I suspect this could also do with a [documentation change](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#finalization-method-dealloc) - probably noting the version dependent behaviour.

> Regarding the tests: Since the PEP was only added into python 3.4, I believe the tests should only run for CI configurations with python >= 3.4. Is there a way to disable the tests for older versions?

Disabling the tests is in https://github.com/cython/cython/blob/4aec02117deb1d9279a81920f21459fc015d1a77/runtests.py#L406

da-woods commented on docs/src/userguide/special_methods.rst:
```diff
@@ -122,7 +122,12 @@ of the superclass will always be called, even if it is overridden.  This is in
 executed unless they are explicitly called by the subclass.
 
-.. Note:: There is no :meth:`__del__` method for extension types.
+Python 3.4 introduced the :meth:`__del__` method to allow for the safe
```
Maybe "Python 3.4 made it possible for extension types to safely define finalizers for objects. When running a Cython module on Python 3.4 and higher you can add..."
then at the end "On Python <3.4 :meth:`__del__` will not be called".

Obviously `__del__` has existing as a Python concept for much longer and so my proposed wording tries to avoid that confusion.

ax487 replied:
OK, I did not know that. Let me just fix that...

ax487 commented:
@da-woods: I added the documentation change and disabled the test for older python versions.

I also took a stab at adapting the code generator to restore the exception state. 

Unfortunately, I as unable to find any documentation on `PyObject_CallFinalizerFromDealloc`, specifically, its return value. I presume that the function returns whether or not the object has been resurrected. As it stands now, the exception state is not restored if the object is resurrected (assuming I presume correctly). Is this correct?

da-woods commented:
Another bug (although this is actually in the existing Cython code):

```
cdef class C:
    def __del__(self):
        print("del C")

cdef class D(C):
    def __del__(self):
        print("del D")
    def __dealloc__(self):
        print("dealloc D")
```

This prints:

```
del D
dealloc D
del D
```

(i.e. `__del__` is called twice). It's not a GC tracked class so there's no inbuilt mechanism for stopping `__del__` being called twice. It gets called first from `__pyx_tp_dealloc_11testdealloc_D`. Then `__pyx_tp_dealloc_11testdealloc_C` and `__del__` gets called again.

It isn't obvious to me right now how to avoid that.

scoder commented:
> It gets called first from `__pyx_tp_dealloc_11testdealloc_D`. Then `__pyx_tp_dealloc_11testdealloc_C` and `__del__` gets called again.

Shouldn't the `FINALIZED` flag set by `PyObject_CallFinalizerFromDealloc()` prevent exactly that?

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1699,6 +1686,31 @@ def generate_dealloc_function(self, scope, code):
+            "if (unlikely("
+            "(PY_VERSION_HEX >= 0x03080000 || __Pyx_PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE))"
+            " && Py_TYPE(o)->tp_finalize) && %s) {" % finalised_check)
+        # if instance was resurrected by finaliser, return
```
This comment belongs to the `return` below.
```suggestion
```

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1699,6 +1686,31 @@ def generate_dealloc_function(self, scope, code):
+            " && Py_TYPE(o)->tp_finalize) && %s) {" % finalised_check)
+        # if instance was resurrected by finaliser, return
+
+        code.putln("{")
+        code.putln("PyObject *etype, *eval, *etb;")
+        code.putln("PyErr_Fetch(&etype, &eval, &etb);")
+        code.putln("const int ressurected = PyObject_CallFinalizerFromDealloc(o);")
+        code.putln("PyErr_Restore(etype, eval, etb);")
+        code.putln("if (ressurected) return;")
+        code.putln("}")
```
C89. And spelling :). And we don't need the extra braces since all of this is inside an `if … {}`
```suggestion
        code.putln("int resurrected;")
        code.putln("PyObject *etype, *eval, *etb;")
        code.putln("PyErr_Fetch(&etype, &eval, &etb);")
        code.putln("resurrected = PyObject_CallFinalizerFromDealloc(o);")
        code.putln("PyErr_Restore(etype, eval, etb);")
        # if instance was resurrected by finaliser, abort finalisation
        code.putln("if (resurrected) return;")
```

da-woods commented:
> I'm not completely sure that putting the code to restore the exception state there is right. I think it should probably happen inside the finalizer (i.e. create a function like `generate_dealloc_function` for `generate_finalizer_function` - it only needs to handle the exception state and nothing else).

Still don't think this is right. I'm suggesting that generating a function like

```
void __pyx_tp_finalizer_classname(PyObject*) {
    // fetch state
    // call "actual" finalizer
    // restore state
}
```

and assigning that function to `tp_finalizer`

The reason I'm suggesting doing it this way is that the PEP says that the exception state should be restored on exit from the finalizer. Doing it outside the finalizer in `__dealloc__` breaks that assumption. Some of this is just about trying to follow the PEP as closely as possible. However I think this breaks with inheritance from non-Cython classes.

Example of a bug that I think is linked

```
cdef class E:
    def __del__(self):
        print("del E")

class F(E):
    def __del__(self):
        print("del F")
        super().__del__()

class G(E):
    pass

# here I just inherit using "true python" rather than Cython
globs = { 'E': E }
exec("""
class H(E):
   pass
""", globs)
H = globs['H']

def func(tp):
    inst = tp()
    raise RuntimeError()
```

`func(F)` works ok, but `func(G)` or `func(H)` report

```
Traceback (most recent call last):
  File "testdealloc.pyx", line 51, in testdealloc.func
    raise RuntimeError()
RuntimeError
Exception ignored in: 'testdealloc.E.__del__'
Traceback (most recent call last):
  File "testdealloc.pyx", line 51, in testdealloc.func
    raise RuntimeError()
RuntimeError: 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <cyfunction func at 0x7f96ef7ccad0> returned NULL without setting an error
```

ax487 commented:
I started on generating the function, but I do not yet understand how to tell the writer to add the function to the `PyMethodDef` block. Could you give me a hand?

da-woods commented:
> I started on generating the function, but I do not yet understand how to tell the writer to add the function to the `PyMethodDef` block. Could you give me a hand?

I've created a PR (onto this PR) that should fill in the missing gaps https://github.com/ax487/cython/pull/1. It hasn't been tested hugely thoroughly but it does fix the exception handling. It may introduce other new issues.

It doesn't fix the "double del for non GC type" bug

da-woods commented:
> Ah, I see. The guard isn't hit because it's not a GC type. Yeah, that's tricky. We'd need to somehow signal to the superclass `tp_dealloc` calls that the finaliser has already run. Maybe we could just set the `FINALIZED` flag ourselves? Because, after all, the object _was_ finalised. Something like that.

I wonder if we could fix this by doing:

```
// in tp_dealloc, around PyObject_CallFinalizerFromDealloc call
if (Py_TYPE(o)->tp_dealloc = cname_of_current_dealloc_function) {
    // consider calling finalizer
}
```

That's basically saying "only call the finalizer from the dealloc of the most derived class". It'd obviously need tests to make sure that it does what I think it should.

ax487 commented:
@da-woods: I added a corresponding check to the generator. The test cases already contain an example with a `cdef class` child of a `cdef class` parent, I added another one with nontrivial `__del__` and `__dealloc__` methods for both parent and child class.

da-woods commented:
That's looking good to me currently thanks. Hopefully the "most derived finalizer" does the right thing.

It'd be good to extend the tests in `test_nontrivial_del_with_exception` to include cases similar to `F` and `G` in https://github.com/cython/cython/pull/3804#issuecomment-686974653 (i.e. inheritance from non-cdef classes. I don't think `H` adds a lot to the test). I believe that they should work correctly, but it's always nice to have tests that prove it.

After that I can't think of too many other corner-cases.

da-woods commented:
To make it work in Python <3.8 you need to add

```
if scope.lookup("__del__"):
            value += "|Py_TPFLAGS_HAVE_FINALIZE"
```

to `TypeFlagsSlot.slot_code` in Cython/Compiler/TypeSlots.py

ax487 commented:
Apparently, some checks are running into a time limit, everything else looks fine. Anything to be done about that?

da-woods commented:
It's an existing issue on MacOS (it's failing to notice that a test has failed so is waiting for it for ever). So nothing to be done here.

ax487 commented:
@da-woods: Is there anything else that needs to be done or can we merge this as it is now?

da-woods commented:
I suspect @scoder will want to have a good look at it before merging. It's the sort of change that it's worth making sure that the behaviour is right, because it's hard to fix later if people end up relying on something unintended. For now I think be patient...

ax487 commented:
No problem, I very much appreciate you help in this.

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1553,6 +1554,27 @@ def generate_new_function(self, scope, code, cclass_entry):
 
+    def generate_del_function(self, scope, code):
+        tp_slot = TypeSlots.ConstructorSlot("tp_finalize", '__del__')
+        slot_func = scope.mangle_internal("tp_finalize")
```
It's enough to do this once.
```suggestion
        slot_func_cname = scope.mangle_internal("tp_finalize")
```

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1553,6 +1554,27 @@ def generate_new_function(self, scope, code, cclass_entry):
             "}")
 
+    def generate_del_function(self, scope, code):
+        tp_slot = TypeSlots.ConstructorSlot("tp_finalize", '__del__')
```
You can use `TypeSlots.get_slot_code_by_name("tp_finalize")` to get the slot code, rather than constructing the `Slot` instance here.

scoder reviewed (commented):
Nice tests.
And I think the `tp_dealloc` check is a good idea, too. It makes sense to expect implementations to call the finaliser themselves.

scoder commented on docs/src/userguide/special_methods.rst:
```diff
@@ -122,7 +122,13 @@ of the superclass will always be called, even if it is overridden.  This is in
+higher you can add a :meth:`__del__` method to extension types in
+order to perform Python cleanup operations. When the :meth:`__del__`
+is called the object is still in a valid state (unlike in the case of
+:meth:`__dealloc___`-ation), permitting the use of Python operations
```
```suggestion
:meth:`__dealloc__`), permitting the use of Python operations
```

scoder commented:
@ax487, I left a couple of suggestions back in September. Would be great if you could try to apply them.

ax487 commented:
Sorry, my brain was still stuck in "awaiting merge" mode.

da-woods commented on Cython/Compiler/TypeSlots.py:
```diff
@@ -503,6 +503,17 @@ def spec_value(self, scope):
+    def slot_code(self, scope):
+        if not scope.lookup_here("__del__"):
+            return 0
+        return InternalMethodSlot.slot_code(self, scope)
```
I'm not 100% sure what's going on with PyPy, but I think the fact this slot is even set is breaking it. I'd change this line to

```
return """
#if CYTHON_USE_TP_FINALIZE
%s
#else
    NULL
#endif
""" % InternalMethodSlot.slot_code(self, scope)
```

and that way avoid setting `tp_finalize` unless we plan to use it

da-woods reviewed (commented):
This looks to be broken on pypy3. I get:

```
cpyext: missing slot wrapper wrap_del
RPython traceback:
  File "pypy_interpreter.c", line 37221, in BuiltinCode_funcrun_obj
  File "pypy_module_cpyext_6.c", line 11925, in wrap_del_call
Fatal RPython error: NotImplementedError
Fatal Python error: Aborted

Stack (most recent call first, approximate line numbers):
  File "<frozen importlib._bootstrap>", line 185 in cb
Aborted (core dumped)
```

To me this looks like something that isn't fixable on Cython's end. I suggest:
1. Add `pep442_tp_finalize` to `tests/pypy_crash_bugs.txt` so the test doesn't get run
2. Make the change suggested in TypeSlots.c to avoid initializing `tp_finalize` unless we plan to use it.

ax487 commented:
I added a guard directly to the slot definition, is that what you meant?

da-woods commented on Cython/Compiler/TypeSlots.py:
```diff
@@ -1021,7 +1021,7 @@ def get_slot_code_by_name(scope, slot_name):
     EmptySlot("tp_del"),
     EmptySlot("tp_version_tag"),
-    DelSlot("tp_finalize", ["__del__"], "0", ifdef="PY_VERSION_HEX >= 0x030400a1"),
+    DelSlot("tp_finalize", ["__del__"], "0", ifdef="PY_VERSION_HEX >= 0x030400a1 && CYTHON_USE_TP_FINALIZE"),
```
I'm don't think that's right. The version I suggested sets the slot to NULL when `CYTHON_USE_TP_FINALIZE` isn't set while your version assumes that the slot doesn't exist.

I admit that my version is really ugly though and there should probably be a better way to do it.

ax487 commented:
@da-woods : You're right, I used your approach instead.

scoder commented on Cython/Compiler/TypeSlots.py:
```diff
@@ -503,6 +503,23 @@ def spec_value(self, scope):
+    def slot_code(self, scope):
+        if not scope.lookup_here("__del__"):
+            return 0
+        return """
+        #if CYTHON_USE_TP_FINALIZE
+        %s
+        #else
+        NULL
+        #endif
+        """ % InternalMethodSlot.slot_code(self, scope)
```
I'd rather not have this part right here, but in a new option similar to `ifdef=`. `slot_code()` should return the plain value, `generate()` takes care of the C-level conditions.

da-woods replied:
@ax487 As a hint (since I think it was me that originally suggested putting this `#if` mess in `slot_code`): https://github.com/cython/cython/blob/f4ce85ba8de0fc26a1e47d1bb6e0027ec95fe24a/Cython/Compiler/TypeSlots.py#L294

da-woods replied:
I've tried to solve this in https://github.com/cython/cython/commit/f7570d7b7bc8b5cac0ae7f567b99a9e3c5826521 (in the interests of getting this change finished).

I don't think I can push it to this PR myself so @scoder could you apply it (if it looks suitable)?

scoder commented on Cython/Compiler/TypeSlots.py:
```diff
@@ -503,6 +503,23 @@ def spec_value(self, scope):
+        super(DelSlot, self).__init__(slot_name, user_methods, default_value, **kwargs)
+
+    def slot_code(self, scope):
+        if not scope.lookup_here("__del__"):
+            return 0
```
Another advantage right here: this can go away (and the whole method, actually), if the C conditions are handled in `generate()`.

scoder reviewed (commented):
Looks mostly good, but the `slot_code()` implementation should be replaced by an option.

da-woods commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1628,6 +1629,27 @@ def generate_new_function(self, scope, code, cclass_entry):
             "}")
 
+    def generate_del_function(self, scope, code):
+        tp_slot = TypeSlots.get_slot_by_name("tp_finalize")
```
```suggestion
        tp_slot = TypeSlots.get_slot_by_name("tp_finalize", scope.directives)
```

(I think).

scoder replied:
Right, that changed in master. Looks reasonable.

da-woods commented on Cython/Compiler/TypeSlots.py:
```diff
@@ -1061,7 +1074,8 @@ def __init__(self, old_binops):
             EmptySlot("tp_version_tag"),
-            EmptySlot("tp_finalize", ifdef="PY_VERSION_HEX >= 0x030400a1"),
+            SyntheticSlot("tp_finalize", ["__del__"], "0", ifdef="PY_VERSION_HEX >= 0x030400a1",
+                          used_ifdef="CYTHON_USE_TP_FINALIZE"),
```
I can't add the suggestion in the right place, but this also needs:

```
MethodSlot(destructor, "", "__del__", method_name_to_slot)
```
to be added just under
```
MethodSlot(destructor, "", "__dealloc__", method_name_to_slot)
```
for `__del__` to be recognised as a special method.

scoder replied:
Right, that was originally added in this PR:
https://github.com/cython/cython/blob/d695dce345cf298c4d660c30c2c64a10737c5593/Cython/Compiler/TypeSlots.py#L1010

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1628,6 +1629,27 @@ def generate_new_function(self, scope, code, cclass_entry):
+        slot_func_cname = scope.mangle_internal("tp_finalize")
+        code.putln("")
+
+        code.putln("#if !CYTHON_COMPILING_IN_LIMITED_API")
```
Shouldn't this condition be aligned with the `used_ifdef` below? That uses `CYTHON_USE_TP_FINALIZE`. However, the latter is also enabled for the limited API. That seems conflicting.

da-woods replied:
Yes I think so. It gives a compile error in the limited API as it is, but works with `#if CYTHON_USE_TP_FINALIZE`

leofang commented:
Would it be possible to backport this PR to the 0.29 series? 🙂

scoder commented:
> Would it be possible to backport this PR to the 0.29 series? slightly_smiling_face

No new features for 0.29.x any more.
